upgrade.py

```diff
@@ -213,9 +222,102 @@ def testUpgradeForbidden(self, tool):
+    def convertTarget(self, progress_callback, primary_disk, target_boot_mode, boot_partnum, primary_partnum, logs_partnum, swap_partnum, storage_partnum, backup_partnum):
+        if not self.safe2mbrupgrade:
+            return
+
+        tool = PartitionTool(primary_disk)
+        if tool.partTableType != constants.PARTITION_DOS:
+            return
+
+        partition_starts = sorted([i['start'] for i in tool.partitions.values()])
+
+        # Verify the primary disk layout is as expected
+        if primary_partnum == 1:
+            # No utility partition
+
+            assert(len(tool.utilityPartitions()) == 0)
+            assert(backup_partnum == 2)
+            assert(storage_partnum == 3)
+
+            assert(len(tool.partitions) == 3)
+        else:
+            # Have utility partition
+            assert(primary_partnum == 2)
+            assert(backup_partnum == 3)
+            assert(storage_partnum == 4)
+
+            assert(len(tool.partitions) == 4)
+
+            # Ensure that utility partition is first on disk
+            assert(tool.partitions[1]['start'] == partition_starts[0])
+
+            # Ensure there is enough space before the utility partition for the GPT
+            assert(tool.partitions[1]['start'] >= constants.gpt_header_512_sectors)
+
+        # Ensure that the SR is the last partition on disk
+        assert(tool.partitions[storage_partnum]['start'] == partition_starts[-1])
+
+        # Ensure new partitions do not exist.
+        boot_partition = tool.getPartition(boot_partnum)
+        assert(boot_partition is None)
+        logs_partition = tool.getPartition(logs_partnum)
+        assert(logs_partition is None)
+        swap_partition = tool.getPartition(swap_partnum)
+        assert(swap_partition is None)
+
+        # This upgrade process is only supported for 512 byte sectors.
+        assert(tool.sectorSize == 512)
+
+        # This upgrade process is only supported for LVM SRs.
+        assert(self.source.storage[0] == diskutil.STORAGE_LVM)
+
+        # Calculate how much space is required at the start and end of the LVM SR.
+        if primary_partnum == 1:
+            existing_install_start = partition_starts[0]
+        else:
+            existing_install_start = partition_starts[1]
+
+        # The new installation start is aligned up to a 1 MiB boundary.
+        # The SR will therefore be resized to a 1 MiB boundary because
+        # the installation is sized in whole MiBs.
+        new_install_start = (existing_install_start + 2047) & ~2047
+        new_sr_start = new_install_start + (constants.install_size * 1024 * 1024 / tool.sectorSize)
+        start_sectors_required = new_sr_start - tool.partitions[storage_partnum]['start']
+        end_sectors_required = max(tool.partitions[storage_partnum]['start'] +
+                                   tool.partitions[storage_partnum]['size'] -
+                                   (tool.sectorExtent - constants.backup_gpt_header_512_sectors), 0)
```

Comment:
I don't understand why we need to reserve an extent at the end of the partition. I mean, I understand there could be an extent that should be moved to make space for the GPT backup, but should not be done here, here we are computing the space for the partition, not the LVM extents.

Also the `max` here looks a bit weird; if the computation would produce a negative number it would look to be a problem, not something to silence.